README
$Id: README,v 1.5 2011/04/22 19:17:10 strauman Exp $

The 'udpComm' API is very similar to traditional socket programming
but slightly simplified.

The 'udpComm' API is implemented for both, communication over regular
BSD sockets and the dedicated 'lanIpBasic' mini-stack.

If you write your application using udpComm then you can later switch
to lanIpBasic w/o modifying your app.

Also, 'udpComm' provides access to lanIpBasic's buffer management
(always for RX, optional for TX) so that copying from/to a user area
can be avoided.

The API is described in

/afs/slac/g/lcls/epics/modules/fcom/devl/src/udpComm/udpComm.h

To use udpComm you need the following:

1) Use rtems-4-9-1-p4 (available under /afs/slac/package/rtems/4.9.1/)
   or later.
   AFAIK, Ernest is copying this release over to the 'builder'.

2) Add a definition for UDPCOMM to your application's RELEASE file.

3) Your application must

      #include <udpComm.h>

4) Add 'udpComm' (when using 'lanIpBasic' -- RTEMS only) or
   'udpCommBSD' (using BSD networking -- supported under RTEMS, linux, ...)
   to the list of libraries used by your application (in your app's Makefile).

   E.g., the BPM application can run on linux (for testing + development,
   i.e., w/o EVR) and on RTEMS (real VME IOC).
   Hence, in my Makefile I say

     bpmApp_LIBS_DEFAULT += udpCommBSD
     bpmApp_LIBS_RTEMS   += udpComm
 

API DESCRIPTION
---------------

NOTE: IP addresses are always 32-bit IPV4 addresses in *network*-byte order
      whereas port numbers are in host-byte order.

SOCKET CREATION
---------------

You must create a 'socket' (in quotes because the referred object is
not a true BSD socket when using lanIpBasic) specifying its UDP port
number (in host-byte order).

  sd = udpCommCreate( 5555 );

JOIN MC GROUP
-------------

Join a MC group (note that the group address is in network byte order)

  udpCommJoinMcast( sd, inet_addr("239.255.3.3") );

(zero return value is success)

PACKET RECEPTION
----------------

Receive/pend for data (a 'udpCommRecvFrom()' variant is also available);
NULL result indicates timeout. A timeout of 0ms instructs to poll,
a negative timeout blocks indefinitely.

  pkt = udpCommRecv( sd, 20000 /*ms*/ );

Find UDP payload in opaque 'packet' (note: it is very important
not to make any assumptions as to the packet layout but to always
use udpCommBufPtr()):

  payld = udpCommBufPtr( pkt );

  /* crunch data */
  printf("Got a string: %s\n", payld);

Multiple users (threads) may share a single packet (read-only). Each
additional 'user' creates an additional reference to the packet

  udpCommRefPacket( pkt );

and MUST release the buffer when done!

  udpCommFreePacket( pkt );

(udpCommFreePacket() decrements the reference count and only truly
releases the packet when the count drops to zero).

LEAVE MC GROUP
--------------

  udpCommLeaveMcast( sd, inet_addr("239.255.3.3") );

(zero result is success)

PACKET TRANSMISSION
-------------------

'connect' socket to 'peer' (which may be a MC group). This is
of course not a true 'connection in the TCP sense but merely
stores the peer's address so that it doesn't have to be
specified with every 'send' operation.

NOTE: the semantics of udpCommConnect() DIFFER from BSD semantics
      if the 'peer' address is a MC group or a broadcast address:
      Connected BSD UDP sockets ONLY accept incoming traffic with
      a source IP address equal to the connected peer's. This means
      that a BSD UDP socket which is connected to a MC group (or
      a broadcast address) will NEVER see any incoming traffic
      (because a packet is never sent *by* a MC group or BC address).

      However, a udpComm socket, when connected to a MC/BC address
      uses the peer information only for sending and will receive
      traffic from any source (IP or port). This applies only if
      the peer IP is a MC group or BC address. In other cases
      (peer is ordinary IP unicast) then BSD semantics are observed.
      
Note that IP addresses are in network- port numbers in host-
byte order.

  udpCommConnect( sd, inet_addr("239.255.3.3"), 6666);

(zero result is success)

Send data to connected peer (user data is copied into network
stack).

  udpCommSend( sd, "Hello\n", 7 );

(positive result is success [# bytes handed to stack])

Alternatively, you can allocate a buffer, fill-in your
data and hand the buffer to the stack (who then 'owns' it
-- you don't release such a buffer ever):

   pkt = udpCommAllocPacket();

   sprintf( udpCommBufPtr(pkt), "Hello\n" );

   udpCommSendPkt( sd, pkt, strlen("Hello\n") + 1 );

(positive result [# of bytes handed to stack] is success)

For some applications this interface allows for avoiding
an extra copy step since lanIpBasic can directly DMA out
of the packet buffer.

SOCKET DESTRUCTION
------------------

   udpCommClose( sd );

ERROR HANDLING
--------------

zero or in some cases a positive return value indicates
SUCCESS. Errors are flagged by negative return values.
An error status is encoded as a 'negative errno', i.e.,
you can get some ASCII info:

   int status = udpCommSomeCall( x, y, z );

   if ( status < 0 ) {
      fprintf( stderr, "Some call failed: %s\n", strerror( - status ) );
   }

MULTIPLE INTERFACES WITH BSD TCP/IP STACK
-----------------------------------------

************************************************************
THIS SECTION ONLY APPLYS IF YOU ARE USING 'TRUE' BSD SOCKETS
************************************************************

Setup
- - -

Make sure you have 'miscUtils.obj' loaded.

1) Attach 2nd NIC to BSD stack:

   ifattach( "mve2", rtems_mve_attach, 0 );

2) Configure 2nd NIC

   ifconf( "mve2", "192.168.0.30", "255.255.255.0" );

Selection of Interface for Multicast
------------------------------------

If you have multiple interfaces then some additional
steps are necessary to ensure that MC traffic goes
across the desired interface.

NOTE: This is only required for udpComm over BSD sockets
(since the BSD stack may have multiple NICs attached).

When using lanIpBasic then the lanIpBasic stack is
totally decoupled from BSD networking and therefore
udpComm implicitly 'knows' which interface to use
(the one attached to lanIpBasic).

However, the facilities described below are also
supported in combination with lanIpBasic to ensure
API compatibility (and to leave it open to add support
for multiple NICs to lanIpBasic).

Transmission
- - - - - - -

Assume you 2nd NIC has
been configured with '192.168.0.30':

   udpCommSetIfMcast( sd, inet_addr("192.168.0.30") );

makes sure all MC traffic sent from 'sd' goes over the
NIC with '192.168.0.30'.

(zero result is success)

Reception
- - - - -

You can set the external variable 'udpCommMcastIfAddr'
to the IP address of the NIC that should receive MC
traffic. You must set this *before* calling udpCommJoinMcast().

NOTE: this is a kludge that was added as a 'quick-n-dirty'
solution. However, the scenario where udpComm over BSD in
combination with multiple NICs is used was deemed rather rare.
(Since usually you would use a single NIC in combination with BSD
-- a second one really doesn't buy you much.)

  udpCommMcastIfAddr = inet_addr("192.168.0.30");
  udpSockJoinMcast( sd, inet_addr("239.255.3.3") );


ChangeLog
20141001 (T.S.)
 - src/drvPadUdpComm.c: lower priority of listener thread during initialization
20140204 (S.H.)
 - src/drvPadUdpComm.c: if epicsTimeGetEvent fails, simply issue warning
   instead of aborting. This change is to support users that will not have
   an event time provider which can handle epicsTimeEventBestTime (SPEAR)
20130905 (T.S.)
 - BUGFIX -- src/udpCommBSD.c: udpCommRecvFrom() returned IP address instead
   of port in *ppeerport.
20130712 (T.S.)
 - src/drvPadUdpComm.c: Made feature query (SQRY) more robust - instead
   of aborting if no digitizer replies delay for 30s and retry (a few times).
   If there is still no answer then make a wild guess.
20130112 (T.S.)
 - Release 2.4 cut
Changes since Release 2.3
20121214 (T.S.)
 - src/drvPadUdpComm.c, src/wavBuf.h, src/wavBuf.c: document semantics
   of 'wavBufPost()': buffer is *always* taken over by this call.
   drvPadUdpComm must not wavBufFree() if wavBufPost() fails.
 - src/drvPadUdpComm.c: BUGFIX: must increase packet reference count
   *before* calling drvPadUdpCommPostRaw.
20121128 (T.S.)
 - src/drvPadUdpComm.c: added iocsh function 'drvPadUdpCommMaxCookReset()'
   for reading and resetting the 'max cook' statistics.
 - src/padProtoHost.c: -m now accepts an optional ': <mcifaddr>' suffix
   so that the interface for incoming multicast traffic can be defined.
 - src/drvPadUdpComm.c, src/padUdpCommSupport.dbd: made 'drvPadUdpCommDebug'
   accessible from iocsh.
 - src/drvPadUdpComm.c: moved consistency checks for stream fragments
   so that they are not performed on the first fragment of a stream (where
   they trivially pass) so that non-fragmented streams incur no overhead.
 - src/drvPadUdpComm.c: keep a copy of the i/o methods in local scope
   (stack) of the Listener. This is in preparation for supporting multiple
   listener threads so that each one can have its own 'cookClosure'.
 - src/drvPadUdpComm.c: BUGFIX - drvPadUdpCommSetupFunc() should only
   return and do nothing if arguments are wrong (had 'return' outside of
   compound instead of inside).
 - src/drvPadUdpComm.h, src/drvPadUdpComm.c, src/hwtime.c, src/Makefile,
   src/fidProcess.c:
    o drvPadUdpCommPktTimeBaseline: now set directly from start command
      (rather than using XID to transport across the channel forth and back).
      Not an per-channel array anymore but just one number.
    - modified drvPadUdpCommHWTime() so that ErGetTicks() can be used
      under the hood. Non-EVR implementation may supply its own version.
      New drvPadUdpCommHWBaseTime() sets 'fidTimeBaseline' (to zero if EVR
      is used).
    - max cook time is now with respect to fidTimeBaseline.
20121120 (T.S.)
 - src/drvPadUdpComm.c:
     o made iocsh string arguments 'persistent' strings.
     o treat empty string args like NULL strings
     o fixed typo reported by Sonya
     o use MAX_BPM (not 1) when scanning for SQRY
  
20121113 (T.S.)
 - src/drvPadUdpComm.c, src/drvPadUdpCommIO.h: added 'int fd' 
   argument to packet allocation function. This allows for supporting
   'per-channel/per-fd' allocators (as needed by the future sis8300 driver).
20121112 (T.S.)
 - src/drvPadUdpComm.c: use 'recv' with indefinite timeout and
   use the watchdog task to issue drvPadUdpCommSendStartAll()
   if no data of any kind on no used channel has arrived.

 - README, README.drvPadUdpComm: updated docs.

 - src/udpCommBSD.c FIX: udpCommRecv timeout < 0 indicates wait forever;
   BSD implementation didn't implement that correctly. Fixed now.

 - src/padStream.h, src/padStream.c, src/padProto.h, src/padProto.c,
   src/hostStream.c, src/drvPadAdcStream.h, src/drvPadAdcStream.c,
   src/drvPadUdpComm.c:
   PADPROTO_VERSION4 adds a 'feature query' command which allows
   the client to find the stream formats supported by the server.

 - src/fidProcess.c: Use macros from event module only if TEST_ONLY
   is not defined.

 - src/drvPadUdpCommIO.h, src/drvPadUdpComm.h, src/drvPadUdpComm.c

   added a new 'setup' callback which is used to negotiate features
   requested by the client and features supported by the server.
   Server features are determined using the new 'stream query'
   protocol command provided by PADPROTO_VERSION4.
   drvPadUdpCommSetup() does not determine the IO methods based
   on the name ("VME" string vs. IP address) anymore but requires
   a new, third argument which passes/defines the IO methods.
   By default, udpComm is used.

 - src/drvPadUdpComm.c: BUGFIX: timeout parameters are passed
   to padRequest in host-byte order.

 - src/drvPadAdcStream.h, src/drvPadAdcStream.c: Moved several
   generic routines supporting the PAD digitizer into separate
   'drvPadAdcStream' module (can be used by multiple apps
   employing PADs).
Changes since Release 2.2
20120524 (T.S.)
 - BUGFIX: src/drvPadAdcStream.c - when reading the FIFOS the first
   couple of samples must be skipped as they generate initial
   clock cycles. Until now we used the stream packet index == 0
   condition to detect the first samples. With multi-packet
   streams this is no longer correct. E.g., in row-major format
   a new row is started at a non-zero index.
   As a result, the second argument to 'drvPadReadFifosXXXMajor()'
   was changed (but is backwards compatible for streams with
   only single packets).
 - API CHANGE: src/padStream.h, src/drvPadAdcStream.h,
   src/drvPadAdcStream.c:
   PadStreamGetdataProc() received a new, additional
   argument: the total number of samples per channel (= 'row dimension').
   This helps the low-level driver with formatting the data.
   CAVEAT: 'getdata' routines on PADs need to be updated to respect
   the new API. 
 - src/drvPadUdpComm.c UDP listener thread priority increased
   from epicsThreadPriorityHigh -> epicsThreadPriorityMax.
 - src/padStream.c: FIXED maximum number of samples accepted into
   a packet. It turns out that the previous value of 1440 would not
   fit (adding all the headers) into a typical MTU of 1500.

src/udpComm.h
/* $Id: udpComm.h,v 1.6 2011/04/20 20:46:44 strauman Exp $ */
#ifndef UDPCOMM_LAYER_H
#define UDPCOMM_LAYER_H

/* Glue for simple UDP communication over
 * either BSD sockets or simple 'udpSock'ets and
 * the lanIpBasic 'stack'.
 */

#include <stdint.h>

#ifdef __cplusplus
extern "C" {
#endif

/* Offset UDPCOMM_DATA_ALGN_OFF into payload
 * is guaranteed to be aligned to UDPCOMM_DATA_ALGN 
 * --- WATCH OUT: payload itself is NOT aligned.
 */
#define UDPCOMM_DATA_ALGN_OFF 4
#define UDPCOMM_DATA_ALGN     16

#ifndef UdpCommPkt
typedef void * UdpCommPkt;
#endif

/* Most routines defined in this interface return a negative
 * error status ('NEGERR') on error whereas a zero or positive
 * return value signals success.
 * The (absolute value of a) negative error code is a standard
 * 'errno' code and may be converted into a string by e.g.,
 *
 * #include <errno.h>
 * #include <string.h>
 *
 *    if ( status < 0 ) fprintf(stderr,"Message: %s\n", strerror(-status));
 */

/* Create; RETURNS: NEGERR (see above) or 0 on success */
int
udpCommSocket(int port);

/* Close; RETURNS: NEGERR (see above) or 0 on success  */
int
udpCommClose(int sd);

/* Connect socket to a peer
 *
 * NOTES: 'dipaddr' is the peer's IP address in *network* byte order
 *        'port'    is the peer's port number in *host*   byte order
 *
 *        If 'dipaddr' is a IP multicast or broadcast address then
 *        BSD 'connect' semantics are not obeyed. Instead, udpComm stores
 *        the destination address/port in memory and uses 'sendto' from
 *        an unconnected socket. (Otherwise the socket would be unable
 *        to receive since only packets from the connected peer [a
 *        MC/BC address in this case] would be accepted).
 * RETURNS: NEGERR (see above) on failure, 0 on success;
 */
int
udpCommConnect(int sd, uint32_t diaddr, int port);

/* Receive a packet; RETURNS: packet handle on success, NULL on failure */
UdpCommPkt
udpCommRecv(int sd, int timeout_ms);

/* Receive a packet and sender information
 *
 * NOTE: port is in host, IP address in network byte order.
 *
 * RETURNS: packet handle on success, NULL on failure.
 */
UdpCommPkt
udpCommRecvFrom(int sd, int timeout_ms, uint32_t *ppeerip, uint16_t *ppeerport);

/* Allocate a packet (for sending with udpCommSendPktTo)
 *
 * RETURNS: packet handle on success, NULL on failure.
 */
UdpCommPkt
udpCommAllocPacket();

/* Release packet (obtained from Recv) when done       */
void
udpCommFreePacket(UdpCommPkt p);

/* Create an additional 'reference' to a packet (increment
 * an internal reference counter).
 * udpCommFreePacket() decrements the reference count and only
 * really releases the packet buffer once the count drops to
 * zero.
 * This facility can be used if you need to re-use the same
 * packet from different software modules.
 * NOTE: Any modifications to the packet data are 'seen' by
 *       ALL users who hold a reference!
 */
void
udpCommRefPacket(UdpCommPkt p);

/* Payload size = eth MTU - ip and udp header sizes    */
#define UDPCOMM_PKTSZ (1500 - 20 - 8)

/* Obtain pointer to data area in buffer (UDP payload) */
void *
udpCommBufPtr(UdpCommPkt p);

/* Send packet to connected peer; 
 * The data in 'buf' has to be copied
 * into the 'lanIpBasic' stack (no-op
 * when using BSD sockets).
 *
 * RETURNS: NEGERR (see above) on failure, number of octets sent on success.
 */
int
udpCommSend(int sd, void *buf, int len);

/* Send packet w/o extra copy step.
 * Packet must be pre-allocated using
 * udpCommAllocPacket() and filled with
 * data (into the user area).
 *
 * NOTE: Ownership of the packet is
 *       transferred to the stack by
 *       this call (regardless of the
 *       return value).
 *
 * RETURNS: NEGERR (see above) on failure, number of octets sent on success.
 */
int
udpCommSendPkt(int sd, UdpCommPkt pkt, int len);

/*
 * As above but send to specified peer
 * (unconnected socket only).
 *
 * NOTE: 'dipaddr' is the peer's IP address in *network* byte order
 *       'port'    is the peer's port number in *host*   byte order
 *
 * RETURNS: NEGERR (see above) on failure, number of octets sent on success.
 */
int
udpCommSendPktTo(int sd, UdpCommPkt pkt, int len, uint32_t dipaddr, int port);

/* Return packet to sender (similar to 'send'; 
 * this interface exists for efficiency reasons
 * [coldfire/lan9118]).
 */
void
udpCommReturnPacket(UdpCommPkt p, int len);

/* Join and leave a MC group. This actually affects the interface
 * not just the socket 'sd'.
 * NOTE: calls do not nest; you cannot call this twice on the
 *       same socket.
 *
 * RETURN: zero on success, NEGERR (see above) on failure.
 */

int
udpCommJoinMcast(int sd, uint32_t mc_addr);

int
udpCommLeaveMcast(int sd, uint32_t mc_addr);

/*
 * Set the outgoing interface for sending multicast
 * traffic from 'sd'. The interface with IP address
 * 'ifaddr' is used for sending all MC traffic from
 * socket 'sd'. 
 *
 * Note that the same or similar functionality 
 * can be achieved by setting up the system routing
 * tables which is more transparent, flexible and
 * IMHO preferable.
 * However, in some cases -- especially for testing --
 * setting this on a 'per-socket' basis with this
 * call is useful; particularly because (on linux
 * and other general-purpose, protected OSes) no
 * special privileges are required.
 *
 * RETURNS: zero on success, NEGERR (see above) on
 *          error.
 *
 * NOTES:   use a 'ifaddr' == 0 to remove the
 *          association of a outgoing mcast IF
 *          with a socket.
 *
 *          The 'ifipaddr' is as usual given in
 *          network-byte order.
 */
int
udpCommSetIfMcastOut(int sd, uint32_t ifipaddr);

/* Set the IP address of the receiving interface
 * (if host has multiple NICs) in network byte order.
 * Defaults to INADDR_ANY, i.e., system picks a suitable IF.
 */
void
udpCommSetIfMcastInp(uint32_t ifipaddr);

#ifdef __cplusplus
}
#endif

#endif
